ouroboros-consensus/src/Ouroboros/Consensus/Storage/ImmutableDB/API.hs

```diff
@@ -1,271 +1,228 @@
+      --
+      -- TODO update
+      --
+      -- Throws an 'InvalidIteratorRangeError' if the start of the range is
+      -- greater than the end of the range.
+      --
+      -- NOTE: 'WrongBoundError' is returned, but 'InvalidIteratorRangeError' is
+      -- thrown. This is because the former is expected to occur during normal
+      -- operation: the user doesn't know upfront if those blocks can be
+      -- streamed or not. Checking it beforehand would be expensive and
+      -- inefficient, and 'streamBlocks' is the best place to do it anyway. The
+      -- latter exception indicates incorrect usage and should not happen during
+      -- normal operation.
+      --
+      -- Throws a 'ReadBlockNewerThanTipError' if the start or end 'SlotNo' are
+      -- in the future.
+      --
+      -- Throws a 'ClosedDBError' if the database is closed.
+      --
+      -- The iterator is automatically closed when exhausted, and can be
+      -- prematurely closed with 'iteratorClose'.
+    , stream_
+        :: forall b. HasCallStack
+        => ResourceRegistry m
+        -> BlockComponent blk b
+        -> StreamFrom blk
+        -> StreamTo   blk
+        -> m (Either (WrongBoundError blk) (Iterator m blk b))
+    }
+  deriving NoUnexpectedThunks via OnlyCheckIsWHNF "ImmutableDB" (ImmutableDB m blk)
+
+{-------------------------------------------------------------------------------
+  Iterator API
+-------------------------------------------------------------------------------}
 
 -- | An 'Iterator' is a handle which can be used to efficiently stream binary
 -- blobs. Slots not containing a blob and missing EBBs are skipped.
-data Iterator hash m a = Iterator
-  { -- | Steps an 'Iterator' yielding an 'IteratorResult'.
-    --
-    -- After returning the blob as an 'IteratorResult', the iterator is
-    -- advanced to the next non-empty slot or non-empty EBB.
-    --
-    -- Throws a 'ClosedDBError' if the database is closed.
-    --
-    -- TODO NOTE(adn): This works under the assumption that the user is the
-    -- exclusive owner of the iterator.
-    --
-    -- The iterator is automatically closed when exhausted
-    -- ('IteratorExhausted'), and can be prematurely closed with
-    -- 'iteratorClose'.
-    iteratorNext    :: HasCallStack => m (IteratorResult a)
-
-    -- | Return the epoch number (in case of an EBB) or slot number and hash
-    -- of the next blob, if there is a next. Return 'Nothing' if not.
-    --
-    -- This operation is idempotent.
-  , iteratorHasNext :: HasCallStack => m (Maybe (Either EpochNo SlotNo, hash))
-
-    -- | Dispose of the 'Iterator' by closing any open handles.
-    --
-    -- Idempotent operation.
-  , iteratorClose   :: HasCallStack => m ()
-  } deriving (Functor)
-
--- | This only contains actions, we don't check anything
-instance NoUnexpectedThunks (Iterator hash m a) where
-  showTypeOf _ = "Iterator"
-  whnfNoUnexpectedThunks _ctxt _itr = return NoUnexpectedThunks
+data Iterator m blk b = Iterator {
```

Comment:
Reuse this in the ChainDB. Maybe instantiate `b` to `Either <GCed> b` in the ChainDB?

Reply:
See #2600.